File: satellites/gameforge-rs/crates/gf-gameforge/src/boardroom.rs
use std::collections::HashMap;
use chrono::{DateTime, Utc};
use serde::{Deserialize, Serialize};
use tokio::sync::RwLock;
use uuid::Uuid;

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
pub enum MotionStatus { Tabled, Resolved, Rejected }

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Motion {
    pub id: String,
    pub session_id: String,
    pub text: String,
    pub mover: String,
    pub tabled_at: DateTime<Utc>,
    pub status: MotionStatus,
    pub resolution_event: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Session {
    pub id: String,
    pub title: String,
    pub agenda: Vec<String>,
    pub opened: DateTime<Utc>,
    pub closed: Option<DateTime<Utc>>,
    pub chair: String,
}

pub struct Boardroom {
    sessions: RwLock<HashMap<String, Session>>,
    motions: RwLock<HashMap<String, Motion>>,
}

impl Default for Boardroom { fn default() -> Self { Self::new() } }
impl Boardroom {
    pub fn new() -> Self { Self { sessions: RwLock::new(HashMap::new()), motions: RwLock::new(HashMap::new()) } }
    pub async fn convene(&self, title: &str, chair: &str, agenda: Vec<String>) -> Session {
        let session = Session { id: Uuid::new_v4().to_string(), title: title.to_string(), agenda, opened: Utc::now(), closed: None, chair: chair.to_string() };
        self.sessions.write().await.insert(session.id.clone(), session.clone());
        session
    }
    pub async fn table_motion(&self, session_id: &str, mover: &str, text: &str) -> Option<(Motion, String)> {
        let sessions = self.sessions.read().await;
        let session = sessions.get(session_id)?;
        if session.closed.is_some() { return None; }
        drop(sessions);
        let motion = Motion { id: Uuid::new_v4().to_string(), session_id: session_id.to_string(), text: text.to_string(), mover: mover.to_string(), tabled_at: Utc::now(), status: MotionStatus::Tabled, resolution_event: None };
        let proposal_id = format!("motion:{}", motion.id);
        self.motions.write().await.insert(motion.id.clone(), motion.clone());
        Some((motion, proposal_id))
    }
    pub async fn resolve_motion(&self, motion_id: &str, carried: bool, event_id: Option<String>) -> Option<Motion> {
        let mut motions = self.motions.write().await;
        let m = motions.get_mut(motion_id)?;
        if m.status != MotionStatus::Tabled { return None; }
        m.status = if carried { MotionStatus::Resolved } else { MotionStatus::Rejected };
        m.resolution_event = event_id;
        Some(m.clone())
    }
    pub async fn adjourn(&self, session_id: &str) -> Option<Session> {
        let mut sessions = self.sessions.write().await;
        let s = sessions.get_mut(session_id)?;
        if s.closed.is_some() { return None; }
        s.closed = Some(Utc::now());
        Some(s.clone())
    }
    pub async fn session_motions(&self, session_id: &str) -> Vec<Motion> {
        self.motions.read().await.values().filter(|m| m.session_id == session_id).cloned().collect()
    }
    pub async fn sessions(&self) -> Vec<Session> { self.sessions.read().await.values().cloned().collect() }
}

File: skeleton/ai/agents/swarm/boardroom.py
"""Boardroom — the inner court's session ledger.

Port of gameforge-rs ``crates/gf-gameforge/src/boardroom.rs``. Where the
zaibatsu's directors convene. A **session** is opened with an agenda,
minutes are minuted as motions, and each motion must cross the high court
before it becomes a **resolution**. Nothing in the boardroom is decided by
presence alone — the quorum gate stands between a motion and its force.
Sessions close with a summary that lands on the fabric: the empire's memory
of what its rulers actually agreed.

``Motion.resolution_event`` holds a fabric / merkle event id once the high
court has spoken (same as RS). Callers that already journal via
``skeleton.kernel.merkle_log.MerkleEventLog`` may pass that append's
``event_id`` into ``resolve_motion`` — extend-only; the merkle log itself
is unchanged.

This is the forge-path session ledger, **not** the file-versioning vault
under ``backend/gameforge/boardroom/``.
"""

from __future__ import annotations

import threading
import time
import uuid
from dataclasses import dataclass, field
from enum import Enum
from typing import Any, Optional

from skeleton.kernel.errors import SkeletonError


class BoardroomError(SkeletonError):
    """Boardroom refuse — closed chamber, already settled, unknown id."""

    code = "SWM.BOARDROOM"
    http_status = 409


class MotionStatus(str, Enum):
    TABLED = "tabled"
    RESOLVED = "resolved"
    REJECTED = "rejected"


@dataclass
class Motion:
    """One tabled minute awaiting (or carrying) the court's verdict."""

    id: str
    session_id: str
    text: str
    mover: str
    tabled_at: float
    status: MotionStatus = MotionStatus.TABLED
    # Fabric / merkle event id once resolved — the motion's provenance.
    resolution_event: Optional[str] = None

    def to_dict(self) -> dict[str, Any]:
        return {
            "id": self.id,
            "session_id": self.session_id,
            "text": self.text,
            "mover": self.mover,
            "tabled_at": self.tabled_at,
            "status": self.status.value,
            "resolution_event": self.resolution_event,
        }


@dataclass
class Session:
    """An open (or adjourned) chamber sitting."""

    id: str
    title: str
    agenda: list[str]
    opened: float
    chair: str
    closed: Optional[float] = None

    def to_dict(self) -> dict[str, Any]:
        return {
            "id": self.id,
            "title": self.title,
            "agenda": list(self.agenda),
            "opened": self.opened,
            "closed": self.closed,
            "chair": self.chair,
        }


class Boardroom:
    """In-process session + motion ledger.

    Thread-safe via a single RLock (mirrors the RS RwLock covering session /
    motion mutation so observers never see a half-written chamber).
    """

    def __init__(self) -> None:
        self._lock = threading.RLock()
        self._sessions: dict[str, Session] = {}
        self._motions: dict[str, Motion] = {}

    def convene(
        self, title: str, chair: str, agenda: list[str] | None = None
    ) -> Session:
        """Open a session with an agenda. Returns the new Session."""
        session = Session(
            id=str(uuid.uuid4()),
            title=title,
            agenda=list(agenda or []),
            opened=time.time(),
            closed=None,
            chair=chair,
        )
        with self._lock:
            self._sessions[session.id] = session
            return Session(
                id=session.id,
                title=session.title,
                agenda=list(session.agenda),
                opened=session.opened,
                closed=session.closed,
                chair=session.chair,
            )

    def table_motion(
        self, session_id: str, mover: str, text: str
    ) -> Optional[tuple[Motion, str]]:
        """Table a motion in an open session.

        Returns ``(motion, proposal_id)`` for the high court to attest on,
        or ``None`` if the session is missing / already closed.
        """
        with self._lock:
            session = self._sessions.get(session_id)
            if session is None or session.closed is not None:
                return None  # no motions in a closed (or ghost) chamber
            motion = Motion(
                id=str(uuid.uuid4()),
                session_id=session_id,
                text=text,
                mover=mover,
                tabled_at=time.time(),
                status=MotionStatus.TABLED,
                resolution_event=None,
            )
            proposal_id = f"motion:{motion.id}"
            self._motions[motion.id] = motion
            return (
                Motion(
                    id=motion.id,
                    session_id=motion.session_id,
                    text=motion.text,
                    mover=motion.mover,
                    tabled_at=motion.tabled_at,
                    status=motion.status,
                    resolution_event=motion.resolution_event,
                ),
                proposal_id,
            )

    def resolve_motion(
        self,
        motion_id: str,
        carried: bool,
        event_id: Optional[str] = None,
    ) -> Optional[Motion]:
        """Record the court's verdict. Called only when the high court has spoken.

        ``event_id`` is the fabric / merkle event id once resolved (RS
        ``resolution_event``). The boardroom never resolves itself.
        """
        with self._lock:
            motion = self._motions.get(motion_id)
            if motion is None or motion.status is not MotionStatus.TABLED:
                return None
            motion.status = (
                MotionStatus.RESOLVED if carried else MotionStatus.REJECTED
            )
            motion.resolution_event = event_id
            return Motion(
                id=motion.id,
                session_id=motion.session_id,
                text=motion.text,
                mover=motion.mover,
                tabled_at=motion.tabled_at,
                status=motion.status,
                resolution_event=motion.resolution_event,
            )

    def adjourn(self, session_id: str) -> Optional[Session]:
        """Close an open session. Returns None if missing or already closed."""
        with self._lock:
            session = self._sessions.get(session_id)
            if session is None or session.closed is not None:
                return None
            session.closed = time.time()
            return Session(
                id=session.id,
                title=session.title,
                agenda=list(session.agenda),
                opened=session.opened,
                closed=session.closed,
                chair=session.chair,
            )

    def session_motions(self, session_id: str) -> list[Motion]:
        with self._lock:
            return [
                Motion(
                    id=m.id,
                    session_id=m.session_id,
                    text=m.text,
                    mover=m.mover,
                    tabled_at=m.tabled_at,
                    status=m.status,
                    resolution_event=m.resolution_event,
                )
                for m in self._motions.values()
                if m.session_id == session_id
            ]

    def sessions(self) -> list[Session]:
        with self._lock:
            return [
                Session(
                    id=s.id,
                    title=s.title,
                    agenda=list(s.agenda),
                    opened=s.opened,
                    closed=s.closed,
                    chair=s.chair,
                )
                for s in self._sessions.values()
            ]
